extensions/ql-vscode/src/query-history.ts

```diff
@@ -884,3 +951,106 @@ the file in the file explorer and dragging it into the workspace.`
     this.treeDataProvider.refresh();
   }
 }
+
+const LAST_SCRUB_TIME_KEY = 'lastScrubTime';
+
+/**
+ * Registers an interval timer that will periodically check for queries old enought
+ * to be deleted.
+ *
+ * Note that this scrubber will clean all queries from all workspaces. It should not
+ * run too often and it should only run from one workspace at a time.
+ *
+ * Generally, `wakeInterval` should be significantly shorter than `throttleTime`.
+ *
+ * @param wakeInterval How often to check to see if the job should run.
+ * @param throttleTime How often to actually run the job.
+ * @param maxQueryTime The maximum age of a query before is ready for deletion.
+ * @param queryDirectory The directory containing all queries.
+ * @param ctx The extension context.
+ */
+export function registerQueryHistoryScubber(
+  wakeInterval: number,
+  throttleTime: number,
+  maxQueryTime: number,
+  queryDirectory: string,
+  ctx: ExtensionContext,
+
+  // optional counter to keep track of how many times the scrubber has run
+  counter?: {
+    increment: () => void;
+  }
+): Disposable {
+  const deregister = setInterval(async () => {
+    const lastScrubTime = ctx.globalState.get<number>(LAST_SCRUB_TIME_KEY);
```

Comment:
Do we also want to store the fact that we're currently scrubbing, in global state, and check that before kicking off the process? That would help avoid multiple workspaces running the scrubber at the same time. We'd need to be careful though about reseting that though.

Or are we happy that if two scrubbing processes did happen to run at the same time, we have enough defensiveness that it wouldn't cause trouble?

Reply:
I think a safe and easy thing to do is to move the `ctx.globalState.update` to happen as soon as possible after the `get`. This minimizes the window when two workspaces would be able to run the scrubber at the same time. Fundamentally though, I don't think this would be a problem. There might be some extra error messages in one of the workspaces if it gets a list of query directories, but by the time it checks for them they are gone.

Reply:
Ok cool - we can see how it goes in practise too and adjust if necessary.